storage/transfermanager/integration_test.go

```diff
@@ -0,0 +1,326 @@
+	}
+
+	w := &testWriter{}
+
+	// Download a single object.
+	d.DownloadObject(ctx, &DownloadObjectInput{
+		Bucket:      bucketName,
+		Object:      objects[0],
+		Destination: w,
+	})
+
+	it := d.Results()
+
+	got, err := it.Next()
+	if err != nil {
+		t.Fatalf("it.Next: %v", err)
+	}
+
+	// Close the writer so we can check the contents.
+	if err := w.Close(); err != nil {
+		t.Fatalf("testWriter.Close: %v", err)
+	}
+
+	if got.Err != nil {
+		t.Errorf("result.Err: %v", got.Err)
+	}
+
+	if got.Object != objects[0] || got.Bucket != bucketName {
+		t.Errorf("expected Bucket(%q).Object(%q), got %q.%q", bucketName, objects[0], got.Bucket, got.Object)
+	}
+
+	if want, got := contentHashes[objects[0]], w.crc32c; got != want {
+		t.Fatalf("content crc32c does not match; got: %v, expected: %v", got, want)
+	}
+
+	if got.Attrs.Size != objectSize {
+		t.Errorf("expected object size %d, got %d", objectSize, got.Attrs.Size)
+	}
+
+	// Download several objects.
+	writers := make([]*testWriter, len(objects))
+	objToWriter := make(map[string]int) // so we can map the resulting content back to the correct object
+	for i, obj := range objects {
+		writers[i] = &testWriter{}
+		objToWriter[obj] = i
+		d.DownloadObject(ctx, &DownloadObjectInput{
+			Bucket:      bucketName,
+			Object:      obj,
+			Destination: writers[i],
+		})
+	}
+
+	if err := d.WaitAndClose(); err != nil {
+		t.Fatalf("d.WaitAndClose: %v", err)
+	}
+
+	// Close the writers so we can check the contents. This should be fine,
+	// since the downloads should all be complete after WaitAndClose.
+	for i := range objects {
+		if err := writers[i].Close(); err != nil {
+			t.Fatalf("testWriter.Close: %v", err)
+		}
+	}
+
+	numResults := 0
+	for {
+		got, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			t.Errorf("it.Next: %v", err)
+		}
+
+		numResults++
+		writerIdx := objToWriter[got.Object]
+
+		if got.Err != nil {
+			t.Errorf("result.Err: %v", got.Err)
+			continue
+		}
+
+		if want, got := contentHashes[got.Object], writers[writerIdx].crc32c; got != want {
+			t.Fatalf("content crc32c does not match; got: %v, expected: %v", got, want)
+		}
+
+		if got.Attrs.Size != objectSize {
+			t.Errorf("expected object size %d, got %d", objectSize, got.Attrs.Size)
+		}
+	}
+
+	if numResults != len(objects) {
+		t.Errorf("expected to receive %d results, got %d results", len(objects), numResults)
+	}
+
+}
+
+// generateRandomFileInGCS uploads a file with random contents to GCS and returns
+// the crc32c hash of the contents.
+func generateFileInGCS(ctx context.Context, o *storage.ObjectHandle, size int64) (uint32, error) {
+	w := o.Retryer(storage.WithPolicy(storage.RetryAlways)).NewWriter(ctx)
+
+	crc32cHash := crc32.New(crc32.MakeTable(crc32.Castagnoli))
+	mw := io.MultiWriter(w, crc32cHash)
+
+	if _, err := io.CopyN(mw, rand.Reader, size); err != nil {
+		w.Close()
+		return 0, err
+	}
+	return crc32cHash.Sum32(), w.Close()
+}
+
+type testWriter struct {
```

Comment:
doesn't have to be in this PR, but we should add a version of this (well, some kind of `DownloaderBuffer` that implements WriterAt) to the library.

Reply:
Yeah, I'd have to look more into that. This is a very barebones implementation that is likely not at all efficient (and doesn't really work as a WriterAt yet).

Reply:
Yeah I think that's fine for now.